pynitrokey/start/upgrade_by_passwd.py

```diff
@@ -427,116 +465,116 @@ def start_update(regnual, gnuk, default_password, password, wait_e, keyno, verbo
 
-    logger.debug('Initial device strings: {}'.format(dev_strings))
+    # @todo: debugging information, log-file only
+    local_print(f"initial device strings: {dev_strings}")
 
     latest_tag = get_latest_release_data()
 
-    local_print('Please note:')
-    local_print('- Latest firmware available is: '
-          '{} (published: {}),\n provided firmware: {}'
-          .format(latest_tag['tag_name'], latest_tag['published_at'], gnuk))
-    local_print('- All data will be removed from the device')
-    local_print('- Do not interrupt the update process, or the device will not run properly')
-    local_print('- Whole process should not take more than 1 minute')
+    local_print(f"Please note:",
+                f"- Latest firmware available is: ",
+                f"  {latest_tag['tag_name']} (published: {latest_tag['published_at']})",
+                f"- provided firmware: {gnuk}",
+                f"- all data will be removed from the device!",
+                f"- do not interrupt update process - the device may not run properly!",
+                f"- the process should not take more than 1 minute")
     if yes:
-        local_print('Accepted automatically')
+        local_print("Accepted automatically")
     else:
-        answer = input('Do you want to continue? [yes/no]: ')
-        local_print('Entered: "{}"'.format(answer))
-        logger.debug('Continue? "{}"'.format(answer))
-        if answer != 'yes':
-            local_print('Device is not modified. Exiting.')
-            exit(1)
+        if not AskUser.strict_yes_no("Do you want to continue?"):
+            local_critical("Exiting due to user request", support_hint=False)
 
     update_done = False
-    for attempt_counter in range(2):
+    retries = 3
+    for attempt_counter in range(retries):
         try:
             # First 4096-byte in data_upgrade is SYS, so, skip it.
             main(wait_e, keyno, passwd, data, data_upgrade[4096:],
                  skip_bootloader, verbosity=verbose)
             update_done = True
             break
+
+        # @todo: add proper exceptions (for each case) here
         except ValueError as e:
-            logger.exception('Error while running update')
-            str_factory_reset = 'Please "factory-reset" your device to ' \
-                                'continue (this will delete all user data from the device) ' \
-                                'and try again with PIN="12345678".'
-            if 'No ICC present' in str(e):
+            local_print("error while running update")
+            str_factory_reset = "Please 'factory-reset' your device to " \
+                "continue (this will delete all user data from the device) " \
+                "and try again with PIN='12345678'"
+
+            if "No ICC present" in str(e):
                 kill_smartcard_services()
-                # local_print('*** Please run update tool again.')
+                local_print("retrying...")
```

Comment:
This needs forced delay between attempts, 2 seconds or so.

Reply:
next PR